### ISSUE TITLE
Screens not loading until music controller is finished loading (which can take a long time)

### ISSUE THREAD

**Issue by Moosty-osu** [ppy/osu#8865]
when i launch lazer it puts me in the game and everything works fine, but i cant open settings or chat and when i go into solo nothing happens just a blank screen, can still go back to the menu once i select solo. if i go into editor it just stays at blank screen with the background but i cant back out unless i try and exit the game then it takes me back to the main menu.

https://streamable.com/q5n6bz

version 2020.418.0 and 2020.421.0

[runtime.log](https://github.com/ppy/osu/files/4533850/runtime.log)
[performance.log](https://github.com/ppy/osu/files/4533874/performance.log)
[debug.log](https://github.com/ppy/osu/files/4533875/debug.log)

**Comment by iiSaLMaN** [comment]
Did the game used to work fine in older versions?

Also, please include the remaining log files. (like `performance.log` and others)

**Comment by Moosty-osu** [comment]
added the other 2 that i think would help, the other ones seemed useless here. the game used to work fine but i just renistalled after a few months, imported my skins just fine and imported beatmaps fine but as soon as i went into solo after the beatmaps were done it just looked like this and when i went back to the main menu it started doing that.

**Comment by iiSaLMaN** [comment]
Does moving the osu!lazer user data folder out of its location fixes this? (located in `%AppData%/osu` for windows or `~/.local/share/osu` for mac and linux)

**Comment by Moosty-osu** [comment]
i dont see a lazer data folder

**Comment by iiSaLMaN** [comment]
Please double-check this again, if it really doesn't exist for you, that means you shouldn't even see a single beatmap after restarting the game as the database file shouldn't even be existing...

Again:
 - On Windows: The folder is `%AppData%/osu`.
 - On macOS / Linux: The folder is `~/.local/share/osu`.

You can move that folder to anywhere you like to let the game create another fresh one for you to test it again with.

**Comment by bdach** [comment]
I think this error is an important clue:
```
2020-04-25 19:20:44 [error]: An unhandled error has occurred.
2020-04-25 19:20:44 [error]: System.InvalidOperationException: BeatmapSets should not be accessed before the music controller is loaded.
2020-04-25 19:20:44 [error]: at osu.Game.Overlays.MusicController.get_BeatmapSets()
2020-04-25 19:20:44 [error]: at osu.Game.Overlays.MusicController.next(Boolean instant)
```
It would seem that the failure scenario is that either the on-screen display or the music controller are loading for a very long time and blocking loads of everything else:

https://github.com/ppy/osu/blob/231a2b9f817d7566117a9f3324212589a789c57b/osu.Game/OsuGame.cs#L599-L601

Can't be anything earlier, as the volume overlay works. OSD's load seems [relatively lightweight](https://github.com/ppy/osu/blob/7eaafe63cb36251b963b876f8f2436acb19f9cad/osu.Game/Overlays/OnScreenDisplay.cs#L50-L55), so by the fact that this seemingly started after a beatmap import I'm guessing that this could be due to the database read in the music controller:

https://github.com/ppy/osu/blob/a890afcac2a4266a5c7ff92fa186f758b166efb5/osu.Game/Overlays/MusicController.cs#L63-L70

and will probably continue to occur even on older versions until the DB is wiped.

@Moosty-osu how many beatmaps did you import, and would you mind attaching `%AppData%/osu/client.db`?

**Comment by NiceAesth** [comment]
Same issue over here when importing my beatmaps from osu!stable (~90k beatmaps). For me it breaks at about 20k beatmaps

**Comment by Moosty-osu** [comment]
@bdach when i started importing it said 34,000 beatmaps, im not sure the exact amount of beatmaps that actually made it but it used up about as much space as stable does. also it says "we dont support that file type" so idk how im supposed to upload the db file, sorry im just not really used to github or any of that

**Comment by bdach** [comment]
Should be able to either zip it or just rename the extension to one of the supported ones. But the beatmap counts mentioned here seem to initially confirm the hypothesis above.

**Comment by Moosty-osu** [comment]
[client.db](https://www.mediafire.com/file/pfgnn8o1pb4h5xy/client.zip/file)
had to upload it to mediafire because it was over 10mb, not sure if thats normal or not, but here

**Comment by Moosty-osu** [comment]
oh hey, it seems to be working now, not sure what i did, didnt move any files, my client.db was still there when i launched, but it looks like it just replaced itself after launching, the file went from like 200mb to like 200kb, so it seems like it got reset. i really have no idea what happened i just woke up and launched the game and it did that. 

https://streamable.com/f8gmqf

if there is anything else you need me to upload that may help figure out what happened or anything like that ill gladly help.

**Comment by bdach** [comment]
I can reproduce with that DB file, and it seems my suspicions to the cause are confirmed. Thanks for providing us all the necessary info!

**Comment by bdach** [comment]
I looked at this for a little while and it's becoming quite clear that EF/sqlite are just absolutely choking on the `GetAllUsableBeatmapSets()` query. At first they block load of the `MusicController` for about a minute or so:

![image](https://user-images.githubusercontent.com/20418176/80304523-e2f92d80-87b6-11ea-83e5-0e58ca8c7f0f.png)

to only then later start blocking the carousel:

![image](https://user-images.githubusercontent.com/20418176/80304537-fefccf00-87b6-11ea-8054-121a0024f743.png)

The fact that the query pulls in the entire beatmapset object graph is not helping. This is the SQL that ends up being generated by SQL for this particular method call:

```sql
SELECT
    "s.Files"."ID",
    "s.Files"."BeatmapSetInfoID",
    "s.Files"."FileInfoID",
    "s.Files"."Filename",
    "b.FileInfo"."ID",
    "b.FileInfo"."Hash",
    "b.FileInfo"."ReferenceCount"
FROM "BeatmapSetFileInfo" AS "s.Files"
INNER JOIN "FileInfo" AS "b.FileInfo"
    ON "s.Files"."FileInfoID" = "b.FileInfo"."ID"
INNER JOIN
    (
        SELECT DISTINCT "s0"."ID"
        FROM "BeatmapSetInfo" AS "s0"
        LEFT JOIN "BeatmapMetadata" AS "s.Metadata0"
            ON "s0"."MetadataID" = "s.Metadata0"."ID"
        WHERE ("s0"."DeletePending" = 0)
            AND ("s0"."Protected" = 0)
    ) AS "t"
    ON "s.Files"."BeatmapSetInfoID" = "t"."ID"
ORDER BY "t"."ID"
```
which for the particular database file attached here generates nearly half a milion records (due to pulling info on *absolutely all* files) and takes about 17 seconds to complete in the sqlite browser application I use.

Not sure what the way forward here is. This might end up requiring major changes to the whole database system, which might be coming with realm anyway; but this is a big consideration to take in moving forward because there are users (not many, but reports like this one show up every now and then) with absolutely huge beatmap collections and to make their experience match stable significant work might be needed to shape all of the systems into place so that accesses can be as efficient as possible.

**Comment by Moosty-osu** [comment]
not sure what any of this osu maker gamer words mean, but just for reference i have imported all my beatmaps to osu twice before, once in like june of last year and once in january. again not sure if that info will help but there it is.

**Comment by Coppertine** [comment]
tl;dr, the possible way to fix the issue is to change how collecting each of the file's information into a table differently... which means alot of work. if not for Realm, but that's like... when completed.

**Comment by peppy** [comment]
The above statement is incorrect. Database structure does not need to change.

**Comment by Coppertine** [comment]
taking from quote of 

> This might end up requiring major changes to the whole database system, which might be coming with realm anyway

**Comment by peppy** [comment]
That is not describing changing structure. Your statement is suggesting this. You did not quote but reworded in a way that changes the meaning.
